python/mp/rosary.py

```diff
@@ -400,56 +553,104 @@ def mainloop(self):
 
 
 
-    @dm.expose()
-    def start(self, interactive=True):
-        """Start the animation loop (aka, mainloop()) and create a shell for live interaction."""
-        r = self
-        if (r.run_mainloop == False):
-            r.run_mainloop = True
-            self.t_mainloop = threading.Thread(name='rosary_mainloop', target=self.mainloop)
-            self.t_mainloop.start()
 
-            # Don't join the thread if being called from server
-            if interactive:
+    ##########################################################################
+    # DEFINE OSC DISPATCHER ROUTES
+    ##########################################################################
+    def fire_trigger(self, trigger_name, *args, **kwargs):
 
-                code.interact(local=locals())
+        requested_trigger = self.trigger_registry.get(trigger_name)
+        # I really only expect there to be one trigger running at a time,
+        # but just in case, get everyone's names
+        running_trigger_names = [t.name for t in self.triggers]
 
-                self.t_mainloop.join()
+        # Don't want to restart a running trigger
+        if requested_trigger is not None and \
+           trigger_name not in running_trigger_names:
 
-    @dm.expose()
-    def stop(self):
-        """Stop the mainloop and exit the application."""
-        self.run_mainloop = False
-        exit(0)
+            # Kill all existing triggers
+            self.clear_triggers()
 
-    @dm.expose()
-    def pause(self):
-        """Stop the animation loop without exiting."""
-        if (self.run_mainloop):
-            self.run_mainloop = False
+            # Start fading out all existing effects
+            self.clear_effects_fade()
 
-    def generate_osc_path(self, fn_name):
+            self.add_trigger_object(requested_trigger(*args, **kwargs))
+
+
+    def turn_knob(self, knob_name, *args, **kwargs):
         """
-        I know this is not consistent with the pattern already established
-        in effect.py, but I feel like just calling /rosary is more intuitive
-        ...for now
+        For all registered effect instance methods mapped to this knob_name,
+        turn the knob (we're assured they're all running effects)
         """
 
-        return "/rosary/{}".format(fn_name)
+        for fn, instance in self.knobs.get(knob_name, []):
+            fn(instance, *args, **kwargs)
+
 
-    def register_with_dispatcher(self):
+    def route_osc_call(self, full_path, *args, **kwargs):
         """
-        Maybe I could move this function to the DispatcherMapper class.
-        Mabye another day.
+        Figure out what the OSC path means and, well, do what the
+        runes instruct us to do.
```

Comment:
This is the magical new method that converts semi-arbitrary (needs to start with /rosary, /effect, or /trigger at least) to kwargs for whatever method follows.